<a href="https://colab.research.google.com/github/shashwatshukla123/class-139/blob/main/Class_139.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install kaggle

In [2]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

!chmod 600 ~/.kaggle/kaggle.json

In [3]:
!kaggle datasets download -d tmdb/tmdb-movie-metadata

 56% 5.00M/8.89M [00:00<00:00, 52.3MB/s]
100% 8.89M/8.89M [00:00<00:00, 56.9MB/s]


In [4]:
!ls

kaggle.json  sample_data  tmdb-movie-metadata.zip


In [5]:
!unzip tmdb-movie-metadata.zip

Archive:  tmdb-movie-metadata.zip
  inflating: tmdb_5000_credits.csv   
  inflating: tmdb_5000_movies.csv    


In [6]:
!ls

kaggle.json  tmdb_5000_credits.csv  tmdb-movie-metadata.zip
sample_data  tmdb_5000_movies.csv


In [7]:
import pandas as pd 
import numpy as np 

df1=pd.read_csv('tmdb_5000_credits.csv')
df2=pd.read_csv('tmdb_5000_movies.csv')

In [8]:
df1.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 4 columns]

In [9]:
df2.head()

budget  ... vote_count
0  237000000  ...      11800
1  300000000  ...       4500
2  245000000  ...       4466
3  250000000  ...       9106
4  260000000  ...       2124

[5 rows x 20 columns]

In [10]:
df1.columns = ['id','tittle','cast','crew']
df2= df2.merge(df1,on='id')

df2.head(5)

budget  ...                                               crew
0  237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1  300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2  245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3  250000000  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4  260000000  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 23 columns]

# Demographic Filtering

Demographics plural :
\
the statistical characteristics of human populations (such as age or income) used especially to identify markets a change in the state's demographics
\
It is refered to a particulat of a poplation.
In demographic filtering we want to offer generalised recomendation to every user.
Step to find out top rated movies.
We nee a metric or scroce to rate movie.
We need to calculate the score for every movie.
Finally we need to sort the score and recommend the best rated movis to user.
Example:- A movie with 8.9 rated and 3 votes,
Another movie with 7.8 rating and 40 votes.
To calculate the best rated among those IMDB [internet movie database] has created a formula it is known as weighted rating and it is famousily used in the industry to get score for their product.




Weighted Rating (WR) = ((v/v+m)R) + ((m/v+m)C)

v - it is the number of votes for a particular movie 
\
m - is the number of votes required to be listed in the chat 
\
R - average rating of the movie 
\
C - mean votes across the whole report.

In [11]:
C = df2['vote_average'].mean()
print("This is the average votes for all movies:-",C)

This is the average votes for all movies:- 6.092171559442011


In [12]:
m = df2['vote_count'].quantile(0.9)
# For a movie to feature in the chat it must have more votes than atleast 90% of the rest of movies in the list.
print(m)

1838.4000000000015


As the output 1838.4 so, if the vote count is more then 1838.4 then we will consider it in chat.

In [13]:
q_movies = df2.copy().loc[df2['vote_count'] >= m]
print(q_movies.shape)

(481, 23)


Copy (copy the content of df2)
\
There are 481 movie whose vote count more then 1838.4 


In [19]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    # v,R  will vary for each movies && m,C  will be same for all the movies.
    #print(R) 
    #print(v)
    #print(m)
    #print(C)
    return (v/(v+m) * R) + (m/(m+v) * C)

q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

We will define a function named as weighted_rating() to create a new column "score" in all the rows.
\
 We are passing individual row, "m" and "C" values to the function. 
 
 Storing the vote count in the "v", storing vote_average in variable "R" we have used apply() function to use the formula because we are dealing with a data frame, and we need to pass individual rows to our function.
 
  Apply() function takes care of that storing the output in "score" column in the data frame.

In [15]:
q_movies = q_movies.sort_values('score', ascending=False)
q_movies[['title', 'vote_count', 'vote_average', 'score']].head(10)

title  ...     score
1881                       The Shawshank Redemption  ...  8.059258
662                                      Fight Club  ...  7.939256
65                                  The Dark Knight  ...  7.920020
3232                                   Pulp Fiction  ...  7.904645
96                                        Inception  ...  7.863239
3337                                  The Godfather  ...  7.851236
95                                     Interstellar  ...  7.809479
809                                    Forrest Gump  ...  7.803188
329   The Lord of the Rings: The Return of the King  ...  7.727243
1990                        The Empire Strikes Back  ...  7.697884

[10 rows x 4 columns]

We have to sort the data frame on the "score" column. 

Sort_values() function of dataframe sorts the values of "score" column asecending=False - creates a descending order we will also output the title, vote_count, vote_average, score (weighted rating) of top 10 movies.

The function head(10) will show the first top 10 rows from the data frame

In [16]:
import plotly.express as px

fig = px.bar((q_movies.head(10).sort_values('score', ascending=True)), x="score", y="title", orientation='h')
fig.show()

### Conclusion:-

Demographic Recommendation provides a general chart of recommended movies to all the users they are not sensitive to the interests of a particular user this is where a more refined system - content based filtering comes into play